corehq/apps/hqadmin/views.py

```diff
@@ -259,6 +262,50 @@ def _to_chart_data(data_dict):
         }
 
 
+@require_superuser_or_developer
+def mass_email(request):
+    if request.method == "POST":
+        form = EmailForm(request.POST)
+        if form.is_valid():
+            subject = form.cleaned_data['email_subject']
+            body_html = Template(form.cleaned_data['email_body_html'])
+            body_text = Template(form.cleaned_data['email_body_text'])
+            real_email = form.cleaned_data['real_email']
+
+            if real_email:
+                recipients = [{
+                    'username': h['username'],
+                    'first_name': h['first_name'] or h['username'],
+                } for h in UserES().web_users().run().hits]
+            else:
+                recipients = [{
+                    'username': request.couch_user.username,
+                    'first_name': request.couch_user.first_name or request.couch_user.username,
+                }]
+
+            for recipient in recipients:
+                text_content = render_to_string("hqadmin/email/mass_email_base.txt", {
+                    'email_body': body_text.render(Context(recipient)),
+                })
+                html_content = render_to_string("hqadmin/email/mass_email_base.html", {
+                    'email_body': body_html.render(Context(recipient)),
+                })
+
+                send_html_email_async.delay(subject, recipient, html_content, text_content,
+                                email_from=settings.DEFAULT_FROM_EMAIL)
```

Comment:
Looks like this is going to keep the very long list of recipients for the arguments to this in rabbit/celery. 
Seeing this a probable cause for high memory usage or storage for the task.
Should we find a way to fetch this in the method called instead of populating it here itself?

Reply:
Agree it's a lot of users (~30K on prod) but not sure what you mean by fetching in the method. Making a wrapper task around `send_html_email_async`?

Reply:
From your comment @mkangia I was thinking that this call was essentially

```
send_html_email_async.delay(subject, [<30K email addresses>], ...)
```
but from the code it looks like
```
for recipient in [<30K email addresses>]:
    send_html_email_async.delay(subject, , ...)
```
Does that change your comment? Or maybe I just misunderstood, and your comment was about having 30K tasks in the celery queue. I'll defer to others on the team as to whether we think creating that many tasks will cause problems. My guess is it won't.

Reply:
@mkangia I'm going to not do this unless you have a reason to believe sending individual tasks will cause a problem. It's not great code, but it is throwaway code. Going to focus on adding error capturing logic instead.

Reply:
@orangejenny Cool. Can skip.